crates/zizmor/src/yaml_patch/mod.rs

```diff
@@ -0,0 +1,1926 @@
+                            {
+                                // Both are mappings, merge them
+                                let mut merged_mapping = existing_mapping.clone();
+                                for (k, v) in new_mapping {
+                                    merged_mapping.insert(k.clone(), v.clone());
+                                }
+
+                                // Use a custom replacement that preserves the key structure
+                                return apply_mapping_replacement(
+                                    content,
+                                    &existing_key_path,
+                                    &key,
+                                    &serde_yaml::Value::Mapping(merged_mapping),
+                                );
+                            }
+                        }
+                    }
+
+                    // Not both mappings, or parsing failed, just replace
+                    return apply_single_operation(
+                        content,
+                        YamlPatchOperation::Replace {
+                            path: existing_key_path,
+                            value: value.clone(),
+                        },
+                    );
+                }
+            }
+
+            // Key doesn't exist, add it using Add operation
+            apply_single_operation(
+                content,
+                YamlPatchOperation::Add {
+                    path: path.clone(),
+                    key: key.clone(),
+                    value: value.clone(),
+                },
+            )
+        }
+
+        YamlPatchOperation::Remove { path } => {
+            if path == "/" {
+                return Err(YamlPatchError::InvalidOperation(
+                    "Cannot remove root document".to_string(),
+                ));
+            }
+            let query = parse_json_pointer_to_query(&path)?;
+            let feature = doc.query(&query)?;
+
+            // For removal, we need to remove the entire line including leading whitespace
+            let start_pos = find_line_start(content, feature.location.byte_span.0);
+            let end_pos = find_line_end(content, feature.location.byte_span.1);
+
+            let mut result = content.to_string();
+            result.replace_range(start_pos..end_pos, "");
+            Ok(result)
+        }
+    }
+}
+
+/// Convert a JSON Pointer path (like "/permissions/contents") to a yamlpath Query
+///
+/// Supports the full JSON Pointer specification (RFC 6901):
+/// - Array indices are automatically detected and parsed as integers
+/// - Special characters are unescaped (~1 becomes /, ~0 becomes ~)
+/// - Empty path components are filtered out
+///
+/// # Examples
+/// - "/permissions/contents" → Query for nested key
+/// - "/jobs/0/steps/1" → Query for array indices
+/// - "/path/with~1slash" → Query for key containing "/"
+pub fn parse_json_pointer_to_query(path: &str) -> Result<yamlpath::Query, YamlPatchError> {
```

Comment:
Architectural question: instead of needing an additional bit of syntax (JSON pointers) for patches here, perhaps we could have patch paths/routes reuse the existing `SymbolicLocation` or `Route` APIs?

If I'm thinking about this right, I _think_ that would allow the "finding" and "fix" layers to share more state/common code here. 

(This is not a hard requirement for this PR or feature, but I'm curious if you have thoughts about it! See https://github.com/zizmorcore/zizmor/issues/779 for some more context as well 🙂)

Reply:
If you already have an idea how to implement this, feel free to edit this PR.